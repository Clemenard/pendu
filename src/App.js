import React, { Component } from 'react'
 import PropTypes from 'prop-types'
import './App.css';


const Letter = ({ index, letter, onClick, usedLetters, disabled }) => (
  <div className={`letter`} onClick={() => onClick(letter, usedLetters)}>
  {  disabled ||
		<button >
      {letter}
    </button>
	}
  </div>
)


Letter.propTypes = {
  letter: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
  onClick: PropTypes.func.isRequired,
	usedLetters: PropTypes.array.isRequired,
	disabled: PropTypes.bool.isRequired,
}





const ALPHABET = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']


class App extends Component {

state = {
	usedLetters : [],
	phrase : '',
  falseCount : 0,
  start : true
}

handlePhraseUpdate = (event) => {

  this.setState({ phrase: event.target.value.toLowerCase()  })
}


computeDisplay= (phrase, usedLetters) =>{
  return phrase.replace(/\w/g,
    (letter) => (usedLetters.includes(letter) ? letter : '_')
  )
}

 addLetter= (letter) =>{
	 	const { usedLetters,falseCount,phrase } = this.state
	usedLetters.push(letter)
  var arrayPhrase= phrase.split('')
  var newCount = arrayPhrase.includes(letter) ? falseCount : falseCount+1
	this.setState({usedLetters: usedLetters})
  this.setState({falseCount: newCount})
}

startHanging = (event) => {
  event.preventDefault()
  this.setState({start: false})
}

restartHanging = (event) => {
  event.preventDefault()
  this.setState({start: true,falseCount :0, usedLetters : [] , phrase :''})
}
render() {
	const { phrase, usedLetters, falseCount,  start } = this.state
	const won = this.computeDisplay(phrase, this.state.usedLetters).indexOf('_')===-1 && phrase.length>4
  const maxFalse = 5
  const lose = (maxFalse-falseCount)<=0

  return (

<div className="pendu ">
{start &&

  <form  onSubmit={this.startHanging}>
        <p>
          <label>
            Entrez le mot mystère
            <input
             type="text"
             id="mystere"
             onChange={this.handlePhraseUpdate}
             value={this.state.phrase} />
          </label>
          <button type="submit">Valider</button>
        </p>
      </form>
}

<p id='pendu'> {this.computeDisplay(phrase, this.state.usedLetters)} </p>
<p> { (maxFalse-falseCount)>0 && !start && 'Il vous reste '+ (maxFalse-falseCount) +' essais'}</p>
<div className='row'>
<div className='col-6'>
  <img src={ 'hang'+(maxFalse-falseCount) +'.png'} alt=""/>
</div>
<div className='col-6 row'>
{
	!start &&
(  ALPHABET.map((letter, index) => (
    <Letter
      letter={letter}
      index={index}
      key={index}
      onClick={this.addLetter}
			usedLetters={usedLetters}
			disabled={usedLetters.includes(letter)}
    />
)))
}
</div>
</div>
<p>{won && 'Gagné'}{lose && 'Perdu'}</p>
{
(won || lose) && <form onSubmit={this.restartHanging}>
  <button type="submit"> Rejouer?</button>
</form>

}
</div>

  )
}}

export default App;
