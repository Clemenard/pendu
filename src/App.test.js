import { expect } from 'chai'
import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'

import App from './App'
import Letter from './App'

describe('<App />', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<App />)
  })
})

it('has 26 letters', () => {
  const wrapper = shallow(<App />)
  expect(wrapper.find('Letter')).to.have.length(26)
})

it('should match its reference snapshot', () => {
  const onClick = sinon.spy()
  const wrapper = shallow(
    <Letter letter="a" feedback="hidden" index={0} onClick={onClick} usedLetter={[]} diasbled />
  )

  expect(wrapper).to.matchSnapshot()
})

it('should match its reference snapshot', () => {
  const wrapper = shallow(<App />)

  expect(wrapper).to.matchSnapshot()
})

describe('<Letter/>', () => {
  it('should trigger its `onClick` prop when clicked', () => {
    const onClick = sinon.spy()
    const wrapper = shallow(
      <Letter letter="a" feedback="hidden" index={0} onClick={onClick} usedLetter={[]} diasbled />
    )

    wrapper.simulate('click')
    expect(onClick).to.have.been.calledWith(0)
  })
})
