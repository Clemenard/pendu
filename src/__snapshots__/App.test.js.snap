// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match its reference snapshot 1`] = `
<div
  className="pendu "
>
  <p
    id="pendu"
  >
     
    _____
     
  </p>
  <div
    className="row"
  >
    <Letter
      disabled={false}
      index={0}
      key="0"
      letter="a"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={1}
      key="1"
      letter="b"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={2}
      key="2"
      letter="c"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={3}
      key="3"
      letter="d"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={4}
      key="4"
      letter="e"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={5}
      key="5"
      letter="f"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={6}
      key="6"
      letter="g"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={7}
      key="7"
      letter="h"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={8}
      key="8"
      letter="i"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={9}
      key="9"
      letter="j"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={10}
      key="10"
      letter="k"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={11}
      key="11"
      letter="l"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={12}
      key="12"
      letter="m"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={13}
      key="13"
      letter="n"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={14}
      key="14"
      letter="o"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={15}
      key="15"
      letter="p"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={16}
      key="16"
      letter="q"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={17}
      key="17"
      letter="r"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={18}
      key="18"
      letter="s"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={19}
      key="19"
      letter="t"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={20}
      key="20"
      letter="u"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={21}
      key="21"
      letter="v"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={22}
      key="22"
      letter="w"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={23}
      key="23"
      letter="x"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={24}
      key="24"
      letter="y"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={25}
      key="25"
      letter="z"
      onClick={[Function]}
      usedLetters={Array []}
    />
  </div>
  <p />
</div>
`;

exports[`should match its reference snapshot 2`] = `
<div
  className="pendu "
>
  <p
    id="pendu"
  >
     
    _____
     
  </p>
  <div
    className="row"
  >
    <Letter
      disabled={false}
      index={0}
      key="0"
      letter="a"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={1}
      key="1"
      letter="b"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={2}
      key="2"
      letter="c"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={3}
      key="3"
      letter="d"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={4}
      key="4"
      letter="e"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={5}
      key="5"
      letter="f"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={6}
      key="6"
      letter="g"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={7}
      key="7"
      letter="h"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={8}
      key="8"
      letter="i"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={9}
      key="9"
      letter="j"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={10}
      key="10"
      letter="k"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={11}
      key="11"
      letter="l"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={12}
      key="12"
      letter="m"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={13}
      key="13"
      letter="n"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={14}
      key="14"
      letter="o"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={15}
      key="15"
      letter="p"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={16}
      key="16"
      letter="q"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={17}
      key="17"
      letter="r"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={18}
      key="18"
      letter="s"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={19}
      key="19"
      letter="t"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={20}
      key="20"
      letter="u"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={21}
      key="21"
      letter="v"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={22}
      key="22"
      letter="w"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={23}
      key="23"
      letter="x"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={24}
      key="24"
      letter="y"
      onClick={[Function]}
      usedLetters={Array []}
    />
    <Letter
      disabled={false}
      index={25}
      key="25"
      letter="z"
      onClick={[Function]}
      usedLetters={Array []}
    />
  </div>
  <p />
</div>
`;
